File: macros/src/PlotTemplate.cc
#include "../include/CentralInclude.h"
#include "TRegexp.h"
#include <sys/stat.h>

using namespace std;

void drawCategory(TString region, TString ptBin);
void setStyle();
TString OutDir = "../Plots/WMass/";
bool scaleQCD=true;
//Rebin mSD
bool RebinMSD=true;
double minMSD=50;
double maxMSD=170;
int N_MSD = 20;
double * new_binning;

// double YMin=0.9*pow(10,1);
// double YMax=pow(10,6);
double YMin=-1;
double YMax=-1;

double xLabelSize=18.0;
double yLabelSize=18.0;
double xTitleSize=20.0;
double yTitleSize=22.0;
double xTitleOffset=4.0;
double yTitleOffset=1.3;


TString InPath=TString(getenv("CMSSW_BASE"))+"/src/UHH2/JetMass/Histograms/W/";
vector<TFile*> BKG_Files={
  new TFile(InPath+"QCD.root","READ"),
  new TFile(InPath+"WMatched.root","READ"),
  new TFile(InPath+"WUnmatched.root","READ")
};
vector<TFile*> SIG_Files={};
vector<TFile*> DAT_Files={new TFile(InPath+"Data.root","READ")};

vector<int> BKG_Colors={867,kOrange-3,kOrange+3};
vector<int> SIG_Colors={};

vector<TString> BKG_Legend={"QCD","W+jets to qq (merged)","W+jets to qq (unmerged)"};
vector<TString> SIG_Legend={};
TString DAT_Legend="Data JetHT 2016";



int main(int argc, char* argv[]){	
  setStyle();
  struct stat buffer;
  if(stat (((std::string) OutDir).c_str(), &buffer) != 0){
    mode_t standard_mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IXGRP | S_IROTH| S_IXOTH;
    mkdir(((std::string) OutDir).c_str(),standard_mode);  
    mkdir(((std::string) (OutDir+"/logY")).c_str(),standard_mode);  
  } 
  double binWidth = (maxMSD-minMSD)/N_MSD;
  vector<double> bins;
  for(int i = 0 ; i<N_MSD ; ++i)bins.push_back(minMSD+i*binWidth);	
  bins.push_back(maxMSD);
  new_binning = &bins[0];
	

  cout << "Plotting Templates from: " << InPath <<endl;
	

  vector<TString> regions = {"_pass","_fail"};
  vector<TString> ptBins = {"","_pt500To550","_pt550To600","_pt600To675","_pt675To800","_pt800To1200"};
  // vector<TString> regions = {"_pass"};
  // vector<TString> ptBins = {""};

  for(auto const region : regions){
    for(auto const ptBin : ptBins){
      drawCategory(region,ptBin);
    }
  }


}


void drawCategory(TString region="pass", TString ptBin=""){
  TString Selection = "N2DDT";
  // TString HistDir = "JetMass_"+Selection+ptBin+"_noJecOnMass"+region;
  TString HistDir = "JetMass_"+Selection+ptBin+region;
  TString HistName = "Mass_central";

  TLegend * legend = new TLegend(0.6,0.70,0.85,0.9);
  legend->SetFillStyle(0);
  legend->SetTextSize(0.03);
  legend->SetMargin(0.4);
  // legend->SetNColumns(2);
  legend->SetColumnSeparation(0.3);
		
  TH1F * dat_Hist = (TH1F*) DAT_Files[0]->Get(HistDir+"/"+HistName);
  for(int i=1;i<DAT_Files.size();++i)dat_Hist->Add((TH1F*) DAT_Files[1]->Get(HistDir+"/"+HistName));

  if(RebinMSD)dat_Hist = (TH1F*) dat_Hist->Rebin(N_MSD,"",new_binning);

  dat_Hist->SetMarkerStyle(8);
  dat_Hist->SetMarkerColor(kBlack);
  dat_Hist->SetLineColor(kBlack);
  legend->AddEntry(dat_Hist,DAT_Legend,"p");

  TH1F * bg_Err_unscaled = (TH1F*) BKG_Files[0]->Get(HistDir+"/"+HistName)->Clone();
  for(int i=1;i<BKG_Files.size();++i)bg_Err_unscaled->Add((TH1F*) BKG_Files[i]->Get(HistDir+"/"+HistName));

  TH1F * qcd_from_data_hist = (TH1F*) dat_Hist->Clone();
  TH1F* qcd_hist = (TH1F*) dat_Hist->Clone();
  qcd_hist->Reset();
  for(int i = 0; i<BKG_Files.size() ; ++i){
    TH1F * bg_Hist =  (TH1F*) BKG_Files[i]->Get(HistDir+"/"+HistName);
    bg_Hist = (TH1F*) bg_Hist->Rebin(N_MSD,"",new_binning);
    if(TString(BKG_Files[i]->GetName()).Contains("QCD")){
      qcd_hist=(TH1F*)bg_Hist->Clone();
      // continue;
    }else{
      qcd_from_data_hist->Add(bg_Hist,-1);
    }
  }
        
  // double QCD_Scale = dat_Hist->Integral()/bg_Err_unscaled->Integral();
  std::cout << "data-w norm: "<<qcd_from_data_hist->Integral() << " qcd norm: " << qcd_hist->Integral() << std::endl;
  double QCD_Scale = qcd_from_data_hist->Integral()/qcd_hist->Integral();
  if(!scaleQCD)QCD_Scale=1.0;

  THStack * bg_Stack = new THStack ("bg_stack","bg_stack");
  TH1F * bg_Err = (TH1F*) BKG_Files[0]->Get(HistDir+"/"+HistName)->Clone();
  if(RebinMSD)bg_Err = (TH1F*) bg_Err->Rebin(N_MSD,"",new_binning);
  bg_Err->Reset();
  for(int i = 0; i<BKG_Files.size() ; ++i){
    TH1F * bg_Hist =  (TH1F*) BKG_Files[i]->Get(HistDir+"/"+HistName);		
    if(RebinMSD)bg_Hist = (TH1F*) bg_Hist->Rebin(N_MSD,"",new_binning);
    if(TString(BKG_Files[i]->GetName()).Contains("QCD")) bg_Hist->Scale(QCD_Scale);
    bg_Hist->SetFillColor(BKG_Colors[i]);
    bg_Hist->SetLineColor(BKG_Colors[i]);
    bg_Stack->Add(bg_Hist);
    legend->AddEntry(bg_Hist,BKG_Legend[i],"f");
    bg_Err->Add(bg_Hist);
  }

			
  bg_Err->SetFillStyle(3204);
  bg_Err->SetFillColor(kGray+2);
  bg_Err->SetLineColor(1);
  legend->AddEntry(bg_Err,"MC stat. Unc.","f");

  bg_Err->GetYaxis()->SetTitle("Events");
  bg_Err->GetYaxis()->SetTitleFont(43);
  bg_Err->GetYaxis()->SetTitleSize(yTitleSize);
  bg_Err->GetYaxis()->SetTitleOffset(yTitleOffset);
  bg_Err->GetYaxis()->SetLabelFont(43);
  bg_Err->GetYaxis()->SetLabelSize(yLabelSize);
  bg_Err->GetXaxis()->SetTitleSize(0.0);
  bg_Err->GetXaxis()->SetLabelSize(0.0);


			
      
  TH1F * ratioHist = (TH1F*) dat_Hist->Clone();
  ratioHist->SetLineColor(kBlack);
  ratioHist->SetStats(0);
  ratioHist->Divide(bg_Err);
  ratioHist->SetMarkerStyle(21);
  ratioHist->SetMarkerSize(0.7);

  ratioHist->GetYaxis()->SetRangeUser(0.45,1.55);
  ratioHist->GetYaxis()->SetTitle("Data/BG");
  ratioHist->GetYaxis()->CenterTitle();
  ratioHist->GetYaxis()->SetTitleFont(43);
  ratioHist->GetYaxis()->SetTitleSize(yTitleSize);
  ratioHist->GetYaxis()->SetTitleOffset(yTitleOffset);
  ratioHist->GetYaxis()->SetLabelFont(43);
  ratioHist->GetYaxis()->SetLabelSize(yLabelSize);
  ratioHist->GetYaxis()->SetNdivisions(506);

  ratioHist->GetXaxis()->SetTitle("m_{SD} [GeV]");
  ratioHist->GetXaxis()->SetTitleFont(43);
  ratioHist->GetXaxis()->SetTitleSize(xTitleSize);
  ratioHist->GetXaxis()->SetTitleOffset(xTitleOffset);
  ratioHist->GetXaxis()->SetLabelFont(43);
  ratioHist->GetXaxis()->SetLabelSize(xLabelSize);
  ratioHist->GetXaxis()->SetTickLength(0.08);
  ratioHist->GetXaxis()->SetNdivisions(506);


  double ratioXMin = ratioHist->GetXaxis()->GetXmin();
  double ratioXMax = ratioHist->GetXaxis()->GetXmax();

  TLine * zeropercent= new TLine(ratioXMin,1,ratioXMax,1);
  TLine * plus10percent= new TLine(ratioXMin,1.1,ratioXMax,1.1);
  TLine * minus10percent= new TLine(ratioXMin,0.9,ratioXMax,0.9);
  plus10percent->SetLineStyle(kDashed);
  minus10percent->SetLineStyle(kDashed);
      
  TCanvas * c1 = new TCanvas("c","c",600,600);
  double yplot=0.7;
  double yratio=0.3;
  double ymax=1.0;
  double xmax=1.0;
  double xmin=0.0;

  int MaxMagnitude=TMath::FloorNint(TMath::Log10(bg_Err->GetMaximum()));
  if(YMin<0){
    YMin=(float) pow(10,MaxMagnitude-3);
    YMin+=(float) pow(10,MaxMagnitude-4);
  }
  if(YMax<0)YMax=(float) pow(10,MaxMagnitude+3);
  // (float) pow(10,MaxMagnitude-5);
  double YMax_lin=bg_Err->GetMaximum()+1*pow(10,MaxMagnitude);
  TPad * plotpad = new TPad("plotpad","Plot",xmin,ymax-yplot,xmax,ymax);
  TPad * ratiopad = new TPad("ratiopad","Ratio",xmin,ymax-yplot-yratio,xmax,ymax-yplot);

  // gROOT->ProcessLine("CMS_lumi(c1);");

  plotpad->SetTopMargin(0.08);
  plotpad->SetLeftMargin(0.1);
  plotpad->SetRightMargin(0.05);
  plotpad->SetTicks();
  plotpad->Draw();

  plotpad->SetBottomMargin(0.016);
  ratiopad->SetTopMargin(0.016);
  ratiopad->SetBottomMargin(0.35);
  ratiopad->SetLeftMargin(0.1);
  ratiopad->SetRightMargin(0.05);
  ratiopad->SetTicks();
  ratiopad->Draw();

  plotpad->cd();
  bg_Err->GetYaxis()->SetRangeUser(0,YMax_lin);
  bg_Err->Draw("E2");
  bg_Stack->Draw("HistSame");
  bg_Err->Draw("E2Same");
  dat_Hist->Draw("PSame");
  plotpad->RedrawAxis();
  TLatex latex = TLatex();
  latex.SetNDC(kTRUE);
  latex.SetTextSize(20);
  latex.DrawLatex(0.69,0.93,TString::Format("%.2f fb^{-1} (13 TeV)",41.86));
  latex.DrawLatex(0.25,0.93,TString::Format("%s N2^{DDT} Selection", region.Contains("p") ? "pass" : "fail" ));

  cout<< "ptBin: " << ptBin <<endl;
  if(ptBin != ""){		
    TString pT_TString_lower=ptBin(TRegexp("[0-9]+To"));
    string pT_string_lower=(string) pT_TString_lower.ReplaceAll("To","");
    double pT_lower = stod(pT_string_lower);
    TString pT_TString_upper=ptBin(TRegexp("To[0-9]+"));
    string pT_string_upper=(string) pT_TString_upper.ReplaceAll("To","");
    double pT_upper = stod(pT_string_upper);
    latex.SetTextSize(25);
    TString ptBin_tex=TString::Format("%.0f GeV #leq p_{T} < %.0f GeV",pT_lower,pT_upper);
    latex.DrawLatex(0.20,0.80,ptBin_tex);
  }


  // latex.DrawLatex(0.1,0.953,"private work");
  legend->Draw("SAME");

	
  ratiopad->cd();
  ratioHist->Draw("EP");
  zeropercent->Draw();
  plus10percent->Draw();
  minus10percent->Draw();

  TString OutName=OutDir+HistName+ptBin+region;				
  c1->SaveAs(OutName+".pdf");

  bg_Err->GetYaxis()->SetRangeUser(YMin,YMax);
  c1->RedrawAxis();
  plotpad->RedrawAxis();
  plotpad->SetLogy();
  c1->SetLogy();
  plotpad->RedrawAxis();
  OutName=OutDir+"logY/"+HistName+ptBin+region;
  c1->SaveAs(OutName+"_logY.pdf");

  delete c1,bg_Stack;
}



void setStyle(){
  //Set TDR styles
  // gROOT->LoadMacro("tdrstyle.C");
  // gROOT->ProcessLine("setTDRStyle();");
	
  // Add CMS text
  // gROOT->LoadMacro("CMS_lumi.C");


  gStyle->SetPadTickY(1);
  gStyle->SetPadTickX(1);
  gStyle->SetLegendBorderSize(0);

  gROOT->SetBatch(true);
  gStyle->SetOptStat(0);
  gStyle->SetOptFit(0);
  gStyle->SetOptTitle(0);

  gStyle->SetTextFont(43);
  //
  gStyle->SetTitleOffset(0.86,"X");
  gStyle->SetTitleOffset(1.6,"Y");
  gStyle->SetPadLeftMargin(0.1);
  gStyle->SetPadBottomMargin(0.12);
  gStyle->SetPadTopMargin(0.08);
  gStyle->SetPadRightMargin(0.1);
  gStyle->SetMarkerSize(0.5);
  gStyle->SetHistLineWidth(1);
  gStyle->SetTitleSize(0.05, "XYZ");
  gStyle->SetLabelSize(0.04, "XYZ");
  gStyle->SetNdivisions(506, "XYZ");
  gStyle->SetLegendBorderSize(0);
}

File: macros/scripts/plot_template.py
import ROOT
from ROOT import gROOT,gStyle
import argparse, os, json
import numpy as np
import sys
sys.path.append(os.environ['CMSSW_BASE']+'/src/UHH2/JetMass/util/python')
import cms_style
cms_style.cms_style()
cms_style.extra_text = "Preliminary"
cms_style.extra_text_rel_X = 0.14
cms_style.font_size_modifier = 0.8
cms_style.text_padding = 0.4


gROOT.SetBatch(True)
gStyle.SetOptStat(0)
gStyle.SetOptFit(0)
gStyle.SetOptTitle(0)

# gStyle.SetTextFont(43)

# gStyle.SetTitleOffset(0.86,"X")
# gStyle.SetTitleOffset(1.8,"Y")
# # gStyle.SetPadLeftMargin(0.18)
# gStyle.SetPadLeftMargin(0.19)
# # gStyle.SetPadBottomMargin(0.15)
# gStyle.SetPadBottomMargin(0.12)
# gStyle.SetPadTopMargin(0.08)
# # gStyle.SetPadRightMargin(0.08)
# gStyle.SetPadRightMargin(0.1)
# gStyle.SetMarkerSize(0.5)
# gStyle.SetHistLineWidth(2)
# gStyle.SetTitleSize(0.05, "XYZ")
# gStyle.SetLabelSize(0.04, "XYZ")
# gStyle.SetNdivisions(506, "XYZ")
# gStyle.SetLegendBorderSize(0)

leftMargin = 0.14



# min_MSD = 0
# max_MSD = 350
# n_MSD = 75
# min_MSD = 0
# max_MSD = 200
# n_MSD = 40
# min_MSD = 40
# max_MSD = 160
# n_MSD = 12

ultrawide = False
legend_on_extern_canvas = False

scaleQCD = True
reverse_stacking = False

# new_binning = np.linspace(min_MSD,max_MSD,n_MSD+1) if n_MSD > 0 else None
#ATLAS binning
#new_binning_dict = {'top':np.linspace(0,350,71),'W':np.linspace(40,160,13),'WfromTop':np.linspace(0,200,40)}
#myBinning
new_binning_dict = {'top':np.linspace(0,350,71),'W':np.linspace(0,500,101),'WfromTop':np.linspace(0,300,61)}
rebin_MSD = True

# rebin_MSD = n_MSD > 0
# print(new_binning)
colors = {
    'QCD':867,
    'QCD_lowPt':867,
    'QCD_highPt':867,
    'WJets':413,
    'WUnmatched':419,
    'WMatched':413,
    'WJetsUnmatched':419,
    'WJetsMatched':413,
    'ZJets':800,
    'DYJets':800,
    'ZJetsUnmatched':797,
    'ZJetsMatched':800,
    'ZUnmatched':797,
    'ZMatched':800,
    'SingleTop':800,
    'ST_tWch':41,
    'ST_tch':43,
    'ST_tW_top':41,
    'ST_tW_antitop':42,
    'ST_tWch_top':41,
    'ST_tWch_antitop':42,
    'ST_tch_top':43,
    'ST_tch_antitop':44,
    'ST_sch':40,    
    'TTbar':810,
    'TTbar_Hadronic':810,
    'TTbar_SemiLeptonic':804,
    'TTToHadronic':810,
    'TTToSemiLeptonic':804,
    'TTbar_had':810,
    'TTbar_semilep':804,
    'TTbar_dilep':803,
    'other':867
}


samples = {
    # "W":["ST_tW_top","ST_tW_antitop","TTbar_SemiLeptonic","TTbar_Hadronic", "ZJetsUnmatched","ZJetsMatched", "WJetsUnmatched","WJetsMatched", "QCD"],
    # "W":["ST_tW_top","ST_tW_antitop","TTToSemiLeptonic","TTToHadronic", "ZJetsUnmatched","ZJetsMatched", "WJetsUnmatched","WJetsMatched","QCD"],
    "W":["ST_tW_top","ST_tW_antitop","TTToSemiLeptonic","TTToHadronic", "ZJetsUnmatched","ZJetsMatched", "WJetsUnmatched","WJetsMatched","QCD"],
    "WfromTop":["QCD", "ST_sch", "ST_tch","ST_tWch", "DYJets", "WJets", "TTbar_had","TTbar_dilep", "TTbar_semilep"],
    "top":["other", "WJets", "SingleTop","TTbar"],
}
pt_bins_dict = {
    "W":["inclusive", "500to550", "550to600", "600to675", "675to800", "800to1200", "1200toInf"],
    "WfromTop":["inclusive", "200to300", "300to400", "400to500", "500toInf"],
    "top":["inclusive", "300to400", "400to500", "500toInf"],
    # "top":["inclusive", "200to300", "300to400", "400toInf"],
}


obs_draw_option = 'PE1'
draw_option = 'H'

ratio_plot = True
xLabelSize=18.
yLabelSize=18.
xTitleSize=20.
yTitleSize=22.
xTitleOffset=2.8
yTitleOffset=1.5

logX=False
logY=False

yplot=0.7
yratio=0.3
ymax=1.0
xmax=1.0
xmin=0.0

y_range = [None,None]
# y_range = [0,2*10**4]
x_range = [None,None]
YRangeUser = all(list(map(lambda a: a is not None ,y_range)))
XRangeUser = all(list(map(lambda a: a is not None ,x_range)))

def setup_hist(hist):
    hist.GetYaxis().SetTitleFont(43)
    hist.GetYaxis().SetTitleSize(yTitleSize)
    hist.GetYaxis().SetTitleOffset(yTitleOffset)
    hist.GetYaxis().SetLabelFont(43)
    hist.GetYaxis().SetLabelSize(yLabelSize)
    if(ratio_plot):
        hist.GetXaxis().SetTitleSize(0.0)
        hist.GetXaxis().SetLabelSize(0.0)
    else:
        hist.GetXaxis().SetTitle(xTitle)
        hist.GetXaxis().SetTitleFont(43)
        hist.GetXaxis().SetTitleSize(xTitleSize)
        hist.GetXaxis().SetTitleOffset(xTitleOffset)
        hist.GetXaxis().SetLabelFont(43)
        hist.GetXaxis().SetLabelSize(xLabelSize)

    if(YRangeUser):
        hist.GetYaxis().SetRangeUser(y_range[0],y_range[1])
    if(XRangeUser):
        hist.GetXaxis().SetRangeUser(x_range[0],x_range[1])
                
def setup_ratio_hist(ratioHist):
  ratioHist.GetYaxis().CenterTitle()
  ratioHist.GetYaxis().SetTitleFont(43)
  ratioHist.GetYaxis().SetTitleSize(yTitleSize)
  ratioHist.GetYaxis().SetTitleOffset(yTitleOffset)
  ratioHist.GetYaxis().SetLabelFont(43)
  ratioHist.GetYaxis().SetLabelSize(yLabelSize)
  ratioHist.GetYaxis().SetNdivisions(506)

  ratioHist.GetXaxis().SetTitle("m_{SD} [GeV]")
  ratioHist.GetXaxis().SetTitleFont(43)
  ratioHist.GetXaxis().SetTitleSize(xTitleSize)
  ratioHist.GetXaxis().SetTitleOffset(xTitleOffset)
  ratioHist.GetXaxis().SetLabelFont(43)
  ratioHist.GetXaxis().SetLabelSize(xLabelSize)
  ratioHist.GetXaxis().SetTickLength(0.08)
  ratioHist.GetXaxis().SetNdivisions(506)

def setup_pads(c):
    if(ratio_plot):
        plotpad = ROOT.TPad("plotpad","Plot",xmin,ymax-yplot,xmax,ymax)
        ratiopad = ROOT.TPad("ratiopad","Ratio",xmin,ymax-yplot-yratio,xmax,ymax-yplot)
    else:
        plotpad = ROOT.TPad("plotpad","Plot",xmin,ymax-yplot-yratio,xmax,ymax)
        ratiopad = None

    c.SetTopMargin(0)
    c.SetBottomMargin(0)
        # plotpad.SetTopMargin(0.1)
    # plotpad.SetLeftMargin(leftMargin)
    # plotpad.SetRightMargin(0.05)
    # plotpad.SetTicks()
    # plotpad.SetTopMargin(0.08)
    # plotpad.SetLeftMargin(leftMargin)
    # plotpad.SetRightMargin(0.05)
    # plotpad.SetTicks()
    plotpad.Draw()
    
    if(ratio_plot):
        plotpad.SetBottomMargin(0.016)
        ratiopad.SetTopMargin(0.016)
        ratiopad.SetBottomMargin(0.35)
        # ratiopad.SetLeftMargin(leftMargin)
        ratiopad.SetRightMargin(0.05)
        ratiopad.SetTicks()
        ratiopad.Draw()
    else:
        plotpad.SetBottomMargin(0.1)
        
    if(logY):
        plotpad.SetLogy()
        c.SetLogy()
    if(logX):
        plotpad.SetLogx()
        if(ratio_plot):
            ratiopad.SetLogx()
        c.SetLogx()
    return plotpad,ratiopad

def plot_template(selection = 'W',hist_file_path = "../Histograms.root",out_dir='../../Plots/Templates',pseudo_data=""):
    f_hists = ROOT.TFile(hist_file_path,"READ")
    ROOT.TH1.AddDirectory(0)

    if(not os.path.exists(out_dir)):
        os.makedirs(out_dir)

    pt_bins = pt_bins_dict[selection]
    pt_bins_str = [(pt_bin.replace('to',' #geq p_{T} > ')+('' if 'in' in pt_bin.lower() else ' GeV')).replace(' > Inf','') for pt_bin in pt_bins]
    
    mc_samples = list(reversed(samples[selection])) if reverse_stacking else samples[selection]
    new_binning = new_binning_dict[selection]

    ext_legend_canvas_done = False
    
    for ipt_bin in range(len(pt_bins)):
        pt_bin = pt_bins[ipt_bin]
        
        for region in ['pass','fail']:
           
            plot_title = '%s %s'%(pt_bins_str[ipt_bin],region)

            c_width = 1800 if ultrawide else 600 
            
            c = ROOT.TCanvas(plot_title,plot_title,c_width,600)            
            legend = ROOT.TLegend(0,0,1,1) if legend_on_extern_canvas else ROOT.TLegend(0.70,0.6,0.9,0.9)
            legend.SetFillStyle(0)
            plotpad,ratiopad = cms_style.setup_pads(c,logY=(logY or selection is 'W'))              
            cms_style.draw_lumi(plotpad, lumi = 41.8,do_extra_text=True,out_of_frame = True,cms_text=True)
            plotpad.cd()
            hist_dir = selection+'_%s__mjet_'+pt_bin+'_'+region
            if(pseudo_data==""):
                h_data = f_hists.Get(hist_dir%"Data")
            else:
                if("nominal" in pseudo_data):
                    hist_dir_pseudo = hist_dir
                else:
                    hist_dir_pseudo = selection+'_%s__mjet_'+pseudo_data[0]+pt_bin+'_'+region+'__'+pseudo_data[1] 
                h_data = f_hists.Get(hist_dir_pseudo%mc_samples[0])
                print(hist_dir_pseudo%mc_samples[0])
                for sample in mc_samples[1:]:
                    print(hist_dir_pseudo%sample)
                    h_data.Add(f_hists.Get(hist_dir_pseudo%sample))
            if(rebin_MSD):
                h_data = h_data.Rebin(len(new_binning)-1,"",new_binning)            
            legend.AddEntry(h_data,'Data','p')

            h_qcd_from_data = h_data.Clone()
            
            bkg_stack = ROOT.THStack()
            bkg_Err = None

            for sample in mc_samples:
                if('QCD' in sample):
                    continue
                this_hist = f_hists.Get(hist_dir%sample)
                if(rebin_MSD):
                    this_hist = this_hist.Rebin(len(new_binning)-1,"",new_binning)                
                    h_qcd_from_data.Add(this_hist,-1)
            legend_mc_entries = []
            for sample in mc_samples:
                this_hist = None
                # if('QCD' in sample):                    
                #     this_hist = f_hists.Get(hist_dir%'QCD_lowPt')
                #     this_hist.Add(f_hists.Get(hist_dir%'QCD_highPt'))
                # else:
                this_hist = f_hists.Get(hist_dir%sample)
                if(rebin_MSD):
                    this_hist = this_hist.Rebin(len(new_binning)-1,"",new_binning)

                #scale QCD to match Data-W-Z (Integral)
                # QCD_scale = 1.0
                if('QCD' in sample and scaleQCD and selection == "W"):
                    norm = this_hist.Integral()
                    QCD_scale = (h_qcd_from_data.Integral()/norm) if norm > 0 else 1.0
                    this_hist.Scale(QCD_scale)
                
                this_hist.SetLineColor(colors[sample])
                this_hist.SetFillColor(colors[sample])
                # legend.AddEntry(this_hist,sample,'f')
                # legend_text = sample + ('* %.2f'%QCD_scale if ('QCD' in sample and QCD_scale != 1.0) else '')
                # legend_text = sample + ('* %.2f'%QCD_scale if ('QCD' in sample and QCD_scale != 1.0) else '')
                legend_mc_entries.append((this_hist,sample,'f'))
                if(bkg_Err is None):
                    bkg_Err = this_hist.Clone()
                else:
                    bkg_Err.Add(this_hist)
                bkg_stack.Add(this_hist,'Hist')
            for entry in reversed(legend_mc_entries):
                legend.AddEntry(*entry)
                
            bkg_Err.SetFillStyle(3204)
            bkg_Err.SetFillColor(922)
            bkg_Err.SetLineWidth(0)
            bkg_Err.SetMarkerSize(0)
            legend.AddEntry(bkg_Err,"MC stat. Unc.","f")

            h_data.SetLineColor(1)
            h_data.SetMarkerStyle(8)
            h_data.SetMarkerSize(0.5)
            h_data.SetTitle(plot_title)
            h_data.Draw(obs_draw_option)

            max_val = max(h_data.GetMaximum(),bkg_Err.GetMaximum())
            h_data.GetYaxis().SetRangeUser(0.9*pow(10,round(np.log10(max_val))-4),1.4*max_val)
            setup_hist(h_data)
            
            bkg_stack.Draw(draw_option + 'SAME')
            bkg_Err.Draw("E2SAME")
            h_data.Draw(obs_draw_option+'SAME')

            if(legend_on_extern_canvas):
                if(not ext_legend_canvas_done):
                    legend_canvas = ROOT.TCanvas('legend','legend',100,150)
                    legend_canvas.cd()
                    legend.Draw()
                    legend_canvas.SaveAs(out_dir+'/legend.pdf')
                    ext_legend_canvas_done=True
                    plotpad.cd()
            else:
                legend.Draw('SAME')

            ratiopad.cd()

            ratio_hist = h_data.Clone()
            ratio_hist.GetYaxis().SetTitle("#frac{Data}{MC}")
            ratio_hist.GetYaxis().SetRangeUser(0.5,1.5)
            ratio_hist.Divide(bkg_Err)
            ratio_hist.SetLineColor(1)
            ratio_hist.SetMarkerColor(1)
            ratio_hist.SetLineWidth(2)
            ratio_hist.SetMarkerStyle(8)
            ratio_hist.SetMarkerSize(0.5)
            setup_ratio_hist(ratio_hist)
            ratio_hist.GetYaxis().SetRangeUser(0.3,1.7)
                

            ratio_hist.Draw('PE1')

            ratioXMin = ratio_hist.GetXaxis().GetXmin()
            ratioXMax = ratio_hist.GetXaxis().GetXmax()

            zeropercent = ROOT.TLine(ratioXMin,1,ratioXMax,1)
            plus10percent = ROOT.TLine(ratioXMin,1.1,ratioXMax,1.1)
            minus10percent = ROOT.TLine(ratioXMin,0.9,ratioXMax,0.9)
            plus10percent.SetLineStyle(ROOT.kDashed)
            minus10percent.SetLineStyle(ROOT.kDashed)
            
            zeropercent.Draw()
            plus10percent.Draw()
            minus10percent.Draw()

                
            plotpad.cd()
            pt_latex = ROOT.TLatex()
            pt_latex.SetNDC(1)
            pt_latex.SetTextSize(cms_style.additional_text_size)
            pt_latex.SetTextFont(42)
            pt_latex.SetTextAlign(31)
            if(pt_bin is not "inclusive"):
                print(pt_bin)
                
                pt_latex.DrawLatex(1-plotpad.GetRightMargin()-0.01,cms_style.additional_text_ypos," %s GeV #leq p_{T} < %s GeV"%(pt_bin.split('to')[0],pt_bin.split('to')[1]))


            plotpad.RedrawAxis()
            c.RedrawAxis()
            c.SaveAs(out_dir+'/'+hist_dir.replace("%s","")+'.pdf')


if(__name__ == '__main__'):
    # plot_template("W","../Histograms.root",'../../Plots/Vqq_jets_templates/')
    # plot_template("W","../scaleStudy/Histograms_oneScale.root",'../../Plots/oneScale_0p01_all_down/',pseudo_data=["0_0_all_","down"])
    # plot_template("W","../Histograms_Cristina.root",'../../Plots/Cristina/templates')
    # plot_template("W","../Histograms.root",'../../Plots/Mine/templates')
    # plot_template("W","../Histograms_Cristina.root",'../../Plots/Cristina/templates_MSD_rawQCD')
    # plot_template("W","../Histograms_Mine.root",'../../Plots/Mine/templates_MSD_rawQCD')
    # plot_template("W","../scaleStudy/Histograms_oneScale.root",'../../Plots/templates_WJets')
    plot_template("W","../Histograms.root",'../../Plots/selection_comparison/wsel')
    # plot_template("WfromTop","../Histograms.root",'../../Plots/selection_comparison/wfromtopsel')
    # plot_template("top","../Histograms_topsel.root",'../../Plots/selection_comparison/topsel')
